### ISSUE TITLE
Data partitioning strategy for horovod

### ISSUE THREAD

**Issue by gismo07** [NVIDIA/DALI#2303]
I'm using horovod to distribute my training process.
However right now it's not 100% clear for me how to handle the partitioning of the dataset. The dataset is present as multiple tfRecord files.

Each worker gets a subset of the tfRecords files to sample the batches:
![nv](https://user-images.githubusercontent.com/27283752/94447737-6970cb80-01aa-11eb-875d-72a950de6661.png)
This mapping however is fixed and consistent throughout the whole training process.

 I've seen, that the `nvidia.dali.ops.TFRecordReader` comes with the two parameters:

- `num_shards ` Partition the data into this many parts (used for multiGPU training).
- `shard_id ` Id of the part to read

These two are also used by the [ tensorflow/resnet-n](https://github.com/NVIDIA/DALI/blob/db2c06f43b33ba0c54a46d66adcc88c43771b109/docs/examples/use_cases/tensorflow/resnet-n/nvutils/image_processing.py) example

Now the question is: is there any shuffling going on by using the DALI shards or is it quite the same approach as the one I'm already using?

I used the hardcoded version because we noticed some heavy slowdowns if we do not pin the workers to different tfRecord files on the same machine (eg. two workers on one node)

**Comment by JanuszL** [comment]
Hi,
DALI threads all tfRecord files as a contiguous data set. Each shard gets a part of it. So in your case, each worker would get a piece of it like in the image below:
![image](https://user-images.githubusercontent.com/39967756/94449734-b786ce80-01ac-11eb-82ae-dd51e7bb51e4.png).
Each epoch each GPU by default move to other shards to increase entropy, but this behavior can be changed by using `stick_to_shard` reader argument.
Shuffling itself is done in the following way:
- create a reasonably big sampling buffer
- read consecutive samples to it
- sample it randomly to form a bach, and replace the picked samples by the one read from the data set

If you want to manually drive how the dataset is shared you need to provide only selected tfRecord and index files to each reader. Also if some file needs to be spitted between shards you need to manually change the index file.

* Index file is a file that keeps the indexes of each entry inside tfRecord files so DALI can easily seek to the first entry in the corresponding shard.

**Comment by gismo07** [comment]
Thank you for the clarification.
I'll try it out immediately.

**Comment by gismo07** [comment]
One question: how does Dali know, when to move to a different shard?

The reader is initialized like this:
```python
reader = ops.TFRecordReader(
            path=tfrFiles,
            index_path=idxFiles,
            features=_f,
            random_shuffle=True,
            preserve=True,
            read_ahead=True,
            num_shards=num_shards,
            shard_id=shard_id,
 )
```

The iterators for pytorch are generated like this:
```python
dali_iter = DALIGenericIterator(
    [pipe], ["images", "steering", "speed"], auto_reset=True
)  # size=epochSize  #TODO: do we need the size arg?
```

and the training loop:
```python
epoch_it = math.ceil(dsLength / BATCH_SIZE)
epoch_it = int(math.ceil(epoch_it / hvd.size()))

for epoch in tqdm(range(EPOCHS)):

    for e in range(epoch_it):

        data = next(dali_iter)

        # train the network

```

**Comment by JanuszL** [comment]
Hi,
The reader knows how many samples are in the dataset and what is the number of shards there is. So when samples are exhausted from the given shard then DALI moves to the next one.

**Comment by gismo07** [comment]
Ok so we dont need to worry about it.
I was wondering, because we don't define the `size` and `reader_name ` for `nvidia.dali.plugin.pytorch.DALIGenericIterator`.

- For `reader_name ` what should be defined here? 
- `size` is the number of samples in the shard for the wrapped pipeline. Is this set automatically?

**Comment by JanuszL** [comment]
Setting `reader_name` is the preferred way.
In this example you may see how to [name the reader](https://github.com/NVIDIA/DALI/blob/61ba11788d7ec2cb4bdd77afb83babe59b951671/docs/examples/use_cases/pytorch/resnet50/main.py#L133) and how to use that name [later on.](https://github.com/NVIDIA/DALI/blob/61ba11788d7ec2cb4bdd77afb83babe59b951671/docs/examples/use_cases/pytorch/resnet50/main.py#L322).

**Comment by JanuszL** [comment]
You can also read more in [the advanced section](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/advanced_topics.html#sharding) of our documentation.

**Comment by JanuszL** [comment]
> I now have the problem that after adding the reader_name property a StopIteration error occurs after the first epoch.

It is expected, then just call reset on the iterator or set `auto_reset` to True and keep iterating.

**Comment by gismo07** [comment]
> > I now have the problem that after adding the reader_name property a StopIteration error occurs after the first epoch.
> 
> It is expected, then just call reset on the iterator or set `auto_reset` to True and keep iterating.

The flag `auto_reset ` was set to True already. So before adding the reader name, the pipeline worked with no problems.

**Comment by JanuszL** [comment]
`reader_name` is mostly intended to properly handle the situation when the data set is not divisible by the number of shard and batch size and the user asks DALI to pad the last batch. You can set `size` to -1 and it will work until the StopIteration is raised from the ExternalSource. If you are not using `ExternalSource` it will work infinitely.
I most cases the user does want to know when the epoch ends and that is why DALI raises StopIteration, and the usual flow is :
```
#training loop
for n in range(number_of_epochs):
  for data in DALITrainIterator():
    # train
  for data in DALIValIterator():
    # validate
```

**Comment by gismo07** [comment]
Ok thanks, so the trick was to set all the `auto_reset` parameters to false and change the training loop to:
```python
for epoch in range(EPOCHS):
    for _e, data in enumerate(dali_iter):
        # do the training
        # count the number of iterations
    
    # reset the iterator
    dali_iter.reset()
```

Now I've noticed, that the total number of iterations per epoch is independent of the number of shards. Can you please explain why? Because according to my logic the number of iterations should decrease with increasing count of shards, because every worker sees only a part of the whole dataset.

**Comment by JanuszL** [comment]
> Now I've noticed, that the total number of iterations per epoch is independent of the number of shards.

By the total number of iterations, I assume you mean the number of elements that `enumerate` returns?
It should not suppose to be like this. Can you provide a simple and self-contained reproduction script of that problem? The best would be to use any data set from https://github.com/NVIDIA/DALI_extra/tree/master/db.

**Comment by gismo07** [comment]
I found the issue. Within our pipeline, we missed to forward the arguments to Dali.
Now it is working like expected. Thank you very much!
